after LLVM 12 update, zig's glibc .so files have unwanted exported symbols

Commit message from 4b1fe8e49238514217e1c6900361a4eb8a0dadbd reproduced here:

glibc: clang 12 assembler regression workaround

Our glibc stub assembly file looked something like this:

```
.globl _Exit_2_2_5
.type _Exit_2_2_5, %function;
.symver _Exit_2_2_5, _Exit@@GLIBC_2.2.5
.hidden _Exit_2_2_5
_Exit_2_2_5:
```

With clang 12, the shared objects this produced stopped having any
exported symbols. When I removed the `.hidden` directive, it resolved
the issue, however, there are now unwanted exports:

```
$ readelf -W --dyn-syms libc.so.6 | grep sys_errlist
   139: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist_GLIBC_2_3
   147: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist_GLIBC_2_4
   395: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist_GLIBC_2_2_5
   487: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist_GLIBC_2_2_5
  1266: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist@@GLIBC_2.12
  1267: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist@GLIBC_2.2.5
  1268: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist@GLIBC_2.3
  1269: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 _sys_errlist@GLIBC_2.4
  2137: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist@@GLIBC_2.12
  2138: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist@GLIBC_2.2.5
  2139: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist@GLIBC_2.3
  2140: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist@GLIBC_2.4
  2156: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist_GLIBC_2_3
  2161: 000000000001ee08     0 FUNC    GLOBAL DEFAULT    7 sys_errlist_GLIBC_2_4
```

Every line here without an `@` symbol is an unwanted export. Before, the
unwanted ones had LOCAL HIDDEN linkage.

As a mitigation, I did two things:

 * Added `_GLIBC_` to the unwanted exports so that they would not
   conflict with anything.
 * Made the default export (the `@@` one) the bare symbol name. This
   appears to reduce the unwanted exports to only symbols that have more
   than one symbol (which is still quite many).

This will unblock progress on this branch, however, there is now a new
issue to solve, that the provided glibc stub .so files have too many
symbols exported. We will have to find a way to avoid this.

Just for the record, I tried to compile Zig with `clang+llvm-12.0.0-rc2-x86_64-linux-gnu-ubuntu-20.10.tar.xz`.
```
/usr/bin/ld: /opt/llvm12/lib/libclang-cpp.so: undefined reference to `tigetnum@NCURSES6_TINFO_5.0.19991023'
/usr/bin/ld: /opt/llvm12/lib/libclang-cpp.so: undefined reference to `del_curterm@NCURSES6_TINFO_5.0.19991023'
/usr/bin/ld: /opt/llvm12/lib/libclang-cpp.so: undefined reference to `set_curterm@NCURSES6_TINFO_5.0.19991023'
/usr/bin/ld: /opt/llvm12/lib/libclang-cpp.so: undefined reference to `setupterm@NCURSES6_TINFO_5.0.19991023'
```

--- a/src/glibc.zig
+++ b/src/glibc.zig
@@ -864,8 +864,8 @@ pub fn buildSharedObjects(comp: *Compilation, prog_node: std.Progress.Node) anye
                     // Example:
                     // .balign 4
                     // .globl _Exit_2_2_5
-                    // .type _Exit_2_2_5, %function;
-                    // .symver _Exit_2_2_5, _Exit@@GLIBC_2.2.5
+                    // .type _Exit_2_2_5, %function
+                    // .symver _Exit_2_2_5, _Exit@@GLIBC_2.2.5, remove
                     // _Exit_2_2_5: .long 0
                     const ver_index = versions_buffer[ver_buf_i];
                     const ver = metadata.all_versions[ver_index];
@@ -876,19 +876,16 @@ pub fn buildSharedObjects(comp: *Compilation, prog_node: std.Progress.Node) anye
                     const want_default = chosen_def_ver_index != 255 and ver_index == chosen_def_ver_index;
                     const at_sign_str: []const u8 = if (want_default) "@@" else "@";
                     if (ver.patch == 0) {
-                        const sym_plus_ver = if (want_default)
-                            sym_name
-                        else
-                            try std.fmt.allocPrint(
-                                arena,
-                                "{s}_GLIBC_{d}_{d}",
-                                .{ sym_name, ver.major, ver.minor },
-                            );
+                        const sym_plus_ver = try std.fmt.allocPrint(
+                            arena,
+                            "{s}_{d}_{d}",
+                            .{ sym_name, ver.major, ver.minor },
+                        );
                         try stubs_asm.writer().print(
                             \\.balign {d}
                             \\.globl {s}
-                            \\.type {s}, %function;
-                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}
+                            \\.type {s}, %function
+                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}, remove
                             \\{s}: {s} 0
                             \\
                         , .{
@@ -904,19 +901,16 @@ pub fn buildSharedObjects(comp: *Compilation, prog_node: std.Progress.Node) anye
                             wordDirective(target),
                         });
                     } else {
-                        const sym_plus_ver = if (want_default)
-                            sym_name
-                        else
-                            try std.fmt.allocPrint(
-                                arena,
-                                "{s}_GLIBC_{d}_{d}_{d}",
-                                .{ sym_name, ver.major, ver.minor, ver.patch },
-                            );
+                        const sym_plus_ver = try std.fmt.allocPrint(
+                            arena,
+                            "{s}_{d}_{d}_{d}",
+                            .{ sym_name, ver.major, ver.minor, ver.patch },
+                        );
                         try stubs_asm.writer().print(
                             \\.balign {d}
                             \\.globl {s}
-                            \\.type {s}, %function;
-                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}.{d}
+                            \\.type {s}, %function
+                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}.{d}, remove
                             \\{s}: {s} 0
                             \\
                         , .{
@@ -1041,9 +1035,9 @@ pub fn buildSharedObjects(comp: *Compilation, prog_node: std.Progress.Node) anye
                     // Example:
                     // .balign 4
                     // .globl environ_2_2_5
-                    // .type environ_2_2_5, %object;
-                    // .size environ_2_2_5, 4;
-                    // .symver environ_2_2_5, environ@@GLIBC_2.2.5
+                    // .type environ_2_2_5, %object
+                    // .size environ_2_2_5, 4
+                    // .symver environ_2_2_5, environ@@GLIBC_2.2.5, remove
                     // environ_2_2_5: .fill 4, 1, 0
                     const ver_index = versions_buffer[ver_buf_i];
                     const ver = metadata.all_versions[ver_index];
@@ -1054,20 +1048,17 @@ pub fn buildSharedObjects(comp: *Compilation, prog_node: std.Progress.Node) anye
                     const want_default = chosen_def_ver_index != 255 and ver_index == chosen_def_ver_index;
                     const at_sign_str: []const u8 = if (want_default) "@@" else "@";
                     if (ver.patch == 0) {
-                        const sym_plus_ver = if (want_default)
-                            sym_name
-                        else
-                            try std.fmt.allocPrint(
-                                arena,
-                                "{s}_GLIBC_{d}_{d}",
-                                .{ sym_name, ver.major, ver.minor },
-                            );
+                        const sym_plus_ver = try std.fmt.allocPrint(
+                            arena,
+                            "{s}_{d}_{d}",
+                            .{ sym_name, ver.major, ver.minor },
+                        );
                         try stubs_asm.writer().print(
                             \\.balign {d}
                             \\.globl {s}
-                            \\.type {s}, %object;
-                            \\.size {s}, {d};
-                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}
+                            \\.type {s}, %object
+                            \\.size {s}, {d}
+                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}, remove
                             \\{s}: .fill {d}, 1, 0
                             \\
                         , .{
@@ -1085,20 +1076,17 @@ pub fn buildSharedObjects(comp: *Compilation, prog_node: std.Progress.Node) anye
                             size,
                         });
                     } else {
-                        const sym_plus_ver = if (want_default)
-                            sym_name
-                        else
-                            try std.fmt.allocPrint(
-                                arena,
-                                "{s}_GLIBC_{d}_{d}_{d}",
-                                .{ sym_name, ver.major, ver.minor, ver.patch },
-                            );
+                        const sym_plus_ver = try std.fmt.allocPrint(
+                            arena,
+                            "{s}_{d}_{d}_{d}",
+                            .{ sym_name, ver.major, ver.minor, ver.patch },
+                        );
                         try stubs_asm.writer().print(
                             \\.balign {d}
                             \\.globl {s}
-                            \\.type {s}, %object;
-                            \\.size {s}, {d};
-                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}.{d}
+                            \\.type {s}, %object
+                            \\.size {s}, {d}
+                            \\.symver {s}, {s}{s}GLIBC_{d}.{d}.{d}, remove
                             \\{s}: .fill {d}, 1, 0
                             \\
                         , .{

